airflow/models.py

```diff
@@ -1020,11 +1020,45 @@ def are_dependents_done(self, session=None):
     @provide_session
     def previous_ti(self, session=None):
         """ The task instance for the task that ran before this task instance """
-        return session.query(TaskInstance).filter(
-            TaskInstance.dag_id == self.dag_id,
-            TaskInstance.task_id == self.task.task_id,
-            TaskInstance.execution_date ==
-            self.task.dag.previous_schedule(self.execution_date),
+        dag = self.task.dag
+
+        if dag.backfill:
+            return session.query(TaskInstance).filter(
+                TaskInstance.dag_id == self.dag_id,
+                TaskInstance.task_id == self.task.task_id,
+                TaskInstance.execution_date ==
+                self.task.dag.previous_schedule(self.execution_date),
+            ).first()
+
+        else:
+            last_dagrun = session.query(DagRun).filter(
+                DagRun.dag_id == dag.dag_id,
+                DagRun.execution_date < self.execution_date
+            ).order_by(
+                DagRun.execution_date.desc()
+            ).first()
+
+            if last_dagrun:
+                return session.query(TaskInstance).filter(
+                    TaskInstance.dag_id == self.dag_id,
+                    TaskInstance.task_id == self.task.task_id,
+                    TaskInstance.execution_date ==
+                    last_dagrun.execution_date,
+                ).first()
+            else:
+                return None
+
+    @property
+    @provide_session
+    def previous_dagrun(self, session=None):
```

Comment:
I put it here, because `ti.previous_dagrun()` seems cleaner than `dr.previous_dagrun(ti)`

It is a helper function, as we could also leverage something like:
```
tip = ti.previous_ti()
if tip:
    drp = tip.get_dagrun()
```